modules/EvseManager/IECStateMachine.cpp

```diff
@@ -0,0 +1,370 @@
+
+    // Process all events
+    while (not events.empty()) {
+        signal_event(events.front());
+        events.pop();
+    }
+}
+
+// Main IEC state machine. Needs to be called whenever:
+// - CP state changes (both events from hardware as well as duty cycle changes)
+// - Allow power on changes
+// - The C1 6s timer expires
+std::queue<CPEvent> IECStateMachine::state_machine() {
+
+    std::queue<CPEvent> events;
+    std::lock_guard lock(state_mutex);
+
+    switch (cp_state) {
+
+    case RawCPState::Disabled:
+        if (last_cp_state != RawCPState::Disabled) {
+            pwm_running = false;
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            timeout_state_c1.stop();
+            call_allow_power_on_bsp(false);
+        }
+        break;
+
+    case RawCPState::A:
+        if (last_cp_state != RawCPState::A) {
+            pwm_running = false;
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            car_plugged_in = false;
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 2.1 Unplug at state Bx (or any other
+        // state) Table A.6: Sequence 2.2 Unplug at state Cx, Dx
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::Disabled) {
+            events.push(CPEvent::CarRequestedStopPower);
+            events.push(CPEvent::CarUnplugged);
+        }
+        break;
+
+    case RawCPState::B:
+        // Table A.6: Sequence 7 EV stops charging
+        // Table A.6: Sequence 8.2 EV supply equipment
+        // responds to EV opens S2 (w/o PWM)
+
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::B) {
+
+            events.push(CPEvent::CarRequestedStopPower);
+            // Need to switch off according to Table A.6 Sequence 8.1
+            // within 100ms
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 1.1 Plug-in
+        if (last_cp_state == RawCPState::A || last_cp_state == RawCPState::Disabled ||
+            (!car_plugged_in && last_cp_state == RawCPState::F)) {
+            events.push(CPEvent::CarPluggedIn);
+            ev_simplified_mode = false;
+        }
+
+        if (last_cp_state == RawCPState::E || last_cp_state == RawCPState::F) {
+            // Triggers SLAC start
+            events.push(CPEvent::EFtoBCD);
+        }
+        break;
+
+    case RawCPState::D:
+        // If state D is not supported switch off.
+        if (not has_ventilation) {
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+            break;
+        }
+        // no break, intended fall through: If we support state D it is handled the same way as state C
+    case RawCPState::C:
+        // Table A.6: Sequence 1.2 Plug-in
+        if (last_cp_state == RawCPState::A || (!car_plugged_in && last_cp_state == RawCPState::F)) {
```

Comment:
What if we go from E > C? We expected the events CarPluggedIn (and CarRequestedPower?) and that we resume charging

Reply:
Hmm technically this is not a plugin on EVSE side, a plugin is always A->B (or C or D). In my understanding E can be used by the EV to signal a (temporary) error. I think at the moment that stops charging and only recovers when you unplug. I can change it so that it restarts charging on E->C if it was authorized before etc.

Reply:
I agree with "this is not a CarPluggedIn event". But Tarragon (EVSE side) is able to generate 0 V, which we need for phase switching. So our question is: can we make it charge again for in this situation and how should the BSP behave in this case?

Reply:
Yes I would change it such that the charging session remains the same (since it is not plugin) but will simply start charging again (similar to the EV pausing within a session). Does that make sense?

Reply:
Yes